Fix OR-NIs event emission + ValueBundle optimized types + custom proto

xmariachi opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v           ✰  Thanks for creating a PR! You're awesome! ✰
v Please note that maintainers will only review those PRs with a completed PR template.
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Purpose of Changes and their Description
* Emit Outlier Resistant network inferences event correctly (using unwrapped context)
* Fix the format, aligning with Network Inferences.

## Are these changes tested and documented?

- [x] If tested, please describe how. If not, why tests are not needed. -- regular unit testing
- [x] If documented, please describe where. If not, describe why docs are not needed. -- no need 
- [x] Added to `Unreleased` section of `CHANGELOG.md`?

xmariachi committed:
Fix OR-NIs event emission + use optimized bundle

xmariachi committed:
remove unused parameters

xmariachi committed:
recreate and clean protos and code

xmariachi committed:
Context alignment

zale144 committed:
small refactor and add important comment

zale144 commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +61,11 @@ message EventNetworkInferences {
 }
 
+message EventOutlierResistantNetworkInferences {
+  reserved 1, 2, 3;
```
why reserve these?

xmariachi replied:
A good practice on evolving protos is to reserve the fields that are no longer used.
Prevents accidental reuse and also makes the intent explicit.

xmariachi replied:
In fact - these ones are evolutions as well, right? They should also reserve the unused field 1 - I can add it to this PR maybe?

```
message EventNetworkLossSet {
  EventValueBundle value_bundle = 2;
}

message EventNetworkInferences {
  EventValueBundle value_bundle = 2;
}
```

zale144 replied:
these should be removed, and in the event emission, we could omit the `v9` bit, since it probably won't make sense to have it anymore

amimart replied:
Maybe it'd make sense to put the topic id at the highest level in the event message instead of the `EventValueBundle`, otherwise we won't be able to filter events on it

zale144 commented on x/emissions/types/events_emitters.go:
```diff
@@ -70,6 +68,21 @@ func EmitNewNetworkInferencesEvent(ctx context.Context, networkInferences ValueB
+		return
+	}
+	sdkCtx.EventManager().EmitEvent(
+		sdk.NewEvent("emissions.v9.EventOutlierResistantNetworkInferences",
```
here I mean, I think it should be just `"emissions.EventOutlierResistantNetworkInferences"`

xmariachi replied:
I think we should still keep the versioning. Future releases may introduce changes. And it makes sense to use the same version as the rest of the events protos.

amimart replied:
I tend to disagree here, the API contract is defined by the proto spec, if not it is not part of that API. It seems very dangerous to me as it can easily be forgotten during a release of a new API version. And it's also misleading, a client could think an associated proto generated type exists and thus use the `cosmostypes.ParseTypedEvent()` func to unmarshal it

xmariachi replied:
But in that case, how would we evolve those types? 
However, I agree on the other concerns - I don't like this "rogue" proto generation if we can avoid it - and it seems like with [your suggestion here](https://github.com/allora-network/allora-chain/pull/859#pullrequestreview-3233635082), it would still be consistent and we can unify and emit typed events in every case, right?

amimart replied:
In that case I would detach the compatibility from the emissions proto API versioning and rely on block height upgrade to identify which event version to use, which is far from ideal...

And yes with that suggestion we can keep all events under the proto API

xmariachi replied:
resolving since with new protos the problem disappears - it will emit typed events only.

guilherme-brandao reviewed (commented):
We need to also check `NewInsertReputerPayloadEventBase`.

amimart commented on x/emissions/types/events_emitters.go:
```diff
@@ -70,6 +68,21 @@ func EmitNewNetworkInferencesEvent(ctx context.Context, networkInferences ValueB
+		sdkCtx.Logger().Warn("Error emitting NewOutlierResistantNetworkInferencesEvent", "error", err)
+		return
+	}
+	sdkCtx.EventManager().EmitEvent(
```
Do we know why the `EventManager#EmitTypedEvent` func doesn't work?

amimart commented on x/emissions/types/events_utils.go:
```diff
@@ -149,6 +150,18 @@ func valueBundleToEventValueBundleBase(bundle *ValueBundle) *EventValueBundle {
+// but in order to emit the event with the actual two-dimensional array which we want, we must use json.Marshal.
+func valueBundleToEventValueBundleJSON(bundle *ValueBundle) ([]byte, error) {
+	evBundle := valueBundleToEventValueBundleBase(bundle)
+	jsn, err := json.Marshal(evBundle)
```
We should avoid using this marshaller as it follows the json rfc 8259 which is not fully interoperable regarding numbers: we may have issues with integers using more than 53bits, which should be encoded as strings.

For the sake of consistency, I think we should use the proto json codec

amimart reviewed (changes requested):
I'm not comfortable with the idea of bypassing the cosmos proto json marshalling, here's what I suggests to still use the default marshalling but having the control only on that double dimensional array:

Let's introduce a custom type for that, implementing the needed proto marshalling funcs:

```go
type DecArray []Dec

var _ sdk.CustomProtobufType = DecArray{}

func (d DecArray) MarshalJSON() ([]byte, error) {
	b := bytes.Buffer{}
	b.WriteByte('[')
	for i, dec := range d {
		if i > 0 {
			b.WriteByte(',')
		}
		decJson, err := dec.MarshalJSON()
		if err != nil {
			return nil, err
		}
		if _, err := b.Write(decJson); err != nil {
			return nil, err
		}
	}
	b.WriteByte(']')
	return b.Bytes(), nil
}

// ...
```

Then in the `EventValueBundle` we can use that custom type:

```proto
repeated string one_out_inferer_forecaster_values = 14 [
    (gogoproto.customtype) = "github.com/allora-network/allora-chain/math.DecArray",
    (gogoproto.nullable) = false
  ];
```

And in the event emission we can keep using the `EventManager#EmitTypedEvent` func

amimart committed:
feat: introduce proto custom type to support 2d dec arrays

amimart committed:
test: add dec array marshalling tests

xmariachi commented:
> We need to also check `NewInsertReputerPayloadEventBase`.

Hey, I think this remark made sense before, but now that we are still emitting typed events, I think it is covered already (it gets the same treatment as the other ones using [valueBundleToEventValueBundleBase](https://github.com/allora-network/allora-chain/blob/3b9ca23a98ac8d644d29ee0f8acd4303312b07a3/x/emissions/types/events_utils.go#L106) e.g. `NewNetworkLossSetEventBase`, `NewNetworkInferencesEventBase`, etc.

cc @amimart

xmariachi commented on math/dec_array_test.go:
```diff
@@ -0,0 +1,173 @@
+			name:      "empty array",
+			data:      []byte(`[]`),
+			expectErr: false,
+			expected:  nil,
```
is empty array expected to be `nil` instead of empty array?

xmariachi replied:
Produced [another PR](https://github.com/allora-network/allora-chain/pull/862) to this branch with additional suggested changes. Let's discuss there

xmariachi replied:
In particular this test has this behaviour, is that expected? 
```
{
			name:      "extra content after array",
			data:      []byte(`["1.0"]extra`),
			expectErr: false,
			expected:  math.DecArray{math.MustNewDecFromString("1.0")},
		},
```

amimart replied:
> is empty array expected to be nil instead of empty array?

Yeah in Go an empty slice should be preferred `nil`, on the encoding side a `nil` slice will always be encoded as `[]`, and `nil` vs empty is the same on the usage, for example:
- `append([]byte(nil), 'x')` == `append([]byte{}, 'x')`
- `len([]byte(nil)) == len([]byte{})`

The test you added should fail, I'll fix that

xmariachi commented on math/dec_array_test.go:
```diff
@@ -0,0 +1,173 @@
+		})
+	}
+}
+
```
`Marshal()` is tested but `MarshalTo()` is not. Even if it is a simple piece of code which does use `Marshal()` , no harm on completing the unit tests for it too, right?

xmariachi replied:
added on [PR](https://github.com/allora-network/allora-chain/pull/862) to this one

xmariachi committed:
edge unit tests

xmariachi committed:
MarshalTo unit tests

xmariachi commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +61,11 @@ message EventNetworkInferences {
   EventValueBundle value_bundle = 2;
 }
```
In v0.12.1, we had these:
```
message EventNetworkLossSet {
  uint64 topic_id = 1;
  int64 block_height = 2;
  emissions.v3.ValueBundle value_bundle = 3;
}

message EventNetworkInferences {
  uint64 topic_id = 1;
  int64 block_height = 2;
  emissions.v3.ValueBundle value_bundle = 3;
}
```
now we are changing it to this:
```
message EventNetworkLossSet {
  EventValueBundle value_bundle = 2;
}

message EventNetworkInferences {
  EventValueBundle value_bundle = 2;
}
```
I don't think this is right: reuses "2", and uses the same name, losing any backwards-compatibility. Also adding your suggestion about keeping topic_id, and to keep compatibility with earlier v0.12.0 and v0.12.1 versions, I would suggest:
```
message EventNetworkLossSet {
  uint64 topic_id = 1;
  int64 block_height = 2; // even if not used now, but it will filled by the prior versions, and can be reserved in v0.13.x
  emissions.v3.ValueBundle value_bundle = 3;  // same here
  EventValueBundle event_value_bundle = 4; // change name, otherwise it may conflict.
}
// ... same for network inferences. 
```
wdyt?

amimart replied:
Agree, I just pushed the related changes marking the old fields as deprecated

xmariachi reviewed (commented):
Some comments, including a PR on unit tests

amimart committed:
fix(math): ensure no extra data after end of decarray when unmarshal

amimart committed:
Diego/engn 4173 decarray edge unit test cases (#862)

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v           ✰  Thanks for creating a PR! You're awesome! ✰
v Please note that maintainers will only review those PRs with a
completed PR template.
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Purpose of Changes and their Description
Add further unit tests to `DecArray`.

## Are these changes tested and documented?

- [x] If tested, please describe how. If not, why tests are not needed.
-- they are unit tests :)
- [x] If documented, please describe where. If not, describe why docs
are not needed. -- no need, just additional unit tests
- [x] Added to `Unreleased` section of `CHANGELOG.md`? -- no need, just
upon another PR.

amimart committed:
fix(x/emissions): ensure evt retro-compatibility

amimart committed:
fix(x/emissions): put the nonce height in network events
